chain-core/src/common/merkle_tree.rs

```diff
@@ -0,0 +1,643 @@
+    path: Path,
+    value: T,
+}
+
+impl<T> Proof<T> {
+    /// Verifies inclusion proof in given merkle tree
+    #[inline]
+    pub fn verify(&self, root_hash: H256) -> bool
+    where
+        T: AsRef<[u8]>,
+    {
+        root_hash == self.root_hash && root_hash == self.path.node_hash && self.path.verify()
+    }
+}
+
+/// Merkle tree with values of type `T`
+#[derive(Debug)]
+pub struct MerkleTree<T> {
+    tree: Tree<T>,
+    len: usize,
+    height: usize,
+}
+
+impl<T> MerkleTree<T> {
+    /// Creates an empty merkle tree
+    #[inline]
+    pub fn empty() -> Self {
+        Self {
+            tree: Tree::empty(),
+            len: 0,
+            height: 0,
+        }
+    }
+
+    /// Creates a new merkle tree with given values as leaves
+    pub fn new(values: Vec<T>) -> Self
+    where
+        T: AsRef<[u8]>,
+    {
+        if values.is_empty() {
+            Self::empty()
+        } else {
+            let len = values.len();
+            let mut trees = values.into_iter().map(Tree::leaf).collect::<Vec<Tree<T>>>();
+            let mut height = 0;
+
+            while trees.len() > 1 {
+                trees = PairIter::from(trees.into_iter())
+                    .map(|(left, right)| match right {
+                        None => left,
+                        Some(right) => Tree::node(Box::new(left), Box::new(right)),
+                    })
+                    .collect::<Vec<Tree<T>>>();
+                height += 1;
+            }
+
+            Self {
+                tree: trees.remove(0),
+                len,
+                height,
+            }
+        }
+    }
+
+    /// Returns root hash of merkle tree
+    #[inline]
+    pub fn root_hash(&self) -> H256 {
+        self.tree.hash()
+    }
+
+    /// Returns height of merkle tree
+    #[inline]
+    pub fn height(&self) -> usize {
+        self.height
+    }
+
+    /// Returns the number of leaf nodes in merkle tree
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Returns `true` if current merkle tree is empty, `false` otherwise
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        0 == self.len
+    }
+
+    /// Generates inclusion proof for given value. Returns `None` if given value is not present in merkle tree
+    #[inline]
+    pub fn generate_proof(&self, value: T) -> Option<Proof<T>>
+    where
+        T: AsRef<[u8]> + Clone,
+    {
+        let root_hash = self.root_hash();
+
+        self.generate_path(value.clone()).map(|path| Proof {
+            root_hash,
+            path,
+            value,
+        })
+    }
+
+    /// Generates merkle path for given value. Returns `None` if given value is not present in merkle tree
+    #[inline]
+    fn generate_path(&self, value: T) -> Option<Path>
+    where
+        T: AsRef<[u8]> + Clone,
+    {
+        self.tree.generate_path(value)
+    }
+}
+
+#[inline]
+fn combine(left: &H256, right: &H256) -> H512 {
+    let mut hash = [0; HASH_SIZE_256 * 2];
```

Comment:
probably not a useful attack in the scenarios how Merkle trees are used here, but it wouldn't hurt to have a fix
https://en.wikipedia.org/wiki/Merkle_tree#Second_preimage_attack
https://flawed.net.nz/2018/02/21/attacking-merkle-trees-with-a-second-preimage-attack/

I guess two ways to fix it:
1) if sparse, then the depth could be stored alongside the root hash and the proof verification would check the proof length against the expected one
2) as suggested in the links: not to construct tree from hashes, but from some hashable / encodable items; prefix encoded leaf items with 0 before hashing them as leafs, intermediate nodes with 1

Reply:
Implemented solution by prefixing `0` (for leaf nodes) or `1` (for intermediate nodes) before hashing.